### PR TITLE
USIO: Added support for Tekken Tag Tournament 2 and Dragon Ball: Zenkai Battle

### PR DESCRIPTION
**Pull request by brian218**
Added support for Tekken Tag Tournament 2.

SCEEXE000_screenshot_2023_07_23
SCEEXE000_screenshot_2023_07_23_1

Support Dragon Ball: Zenkai Battle by the same function as well.
SCEEXE000_screenshot_2023_07_23_2

With this PR merged, the following 3 System 357/369 arcade games are fully supported by RPCS3:
 - Taiko no Tatsujin series
 - Tekken Tag Tournament 2 (Unlimited)
 - Dragon Ball: Zenkai Battle

By the way, in my opinion, there's no any point keeping the small drum hits for Taiko, so I removed them to make the configuration tidy. It's not by accident.

### CODE REVIEW

File: rpcs3/Emu/Cell/lv2/sys_usbd.cpp [hunk 1]
```diff
@@ -471,7 +471,7 @@ void usb_handler_thread::operator()()
 		if (handled_devices.empty())
 			thread_ctrl::wait_for(500'000);
 		else
-			thread_ctrl::wait_for(1'000);
+			thread_ctrl::wait_for(16);
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_usbd.cpp:474]: This is microseconds not miliseconds.

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_usbd.cpp]: Sorry, my mistake.

File: rpcs3/Emu/Io/usio.cpp [hunk 2]
```diff
@@ -401,44 +475,50 @@ void usb_device_usio::usio_read(u8 channel, u16 reg, u16 size)
...
+			// "NBGI.;USIO01;Ver1.00;JPN,Multipurpose with PPG."
+			response = {0x4E, 0x42, 0x47, 0x49, 0x2E, 0x3B, 0x55, 0x53, 0x49, 0x4F, 0x30, 0x31, 0x3B, 0x56, 0x65, 0x72, 0x31, 0x2E, 0x30, 0x30, 0x3B, 0x4A, 0x50, 0x4E, 0x2C, 0x4D, 0x75, 0x6C, 0x74, 0x69, 0x70, 0x75, 0x72, 0x70, 0x6F, 0x73, 0x65, 0x20, 0x77, 0x69, 0x74, 0x68, 0x20, 0x50, 0x50, 0x47, 0x2E, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x13, 0x00, 0x30, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x03, 0x02, 0x00, 0x08, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03, 0x00, 0x03, 0x00, 0x75, 0x6C, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
 			break;
 		}
 		default:
 		{
-			usio_log.error("Unhandled channel 0 register read: 0x%04X", reg);
+			//usio_log.error("Unhandled channel 0 register read: 0x%04X", reg);
```
Comment (Megamouse) [inline rpcs3/Emu/Io/usio.cpp:507]: Either remove or set to trace

Comment (elad335) [inline rpcs3/Emu/Io/usio.cpp:507]: Why this change?

File: rpcs3/Emu/Io/usio.cpp [hunk 3]
```diff
@@ -221,89 +222,162 @@ void usb_device_usio::translate_input()
...
 
-	test_keys |= test_on ? 0x80 : 0x00;
-	std::memcpy(input_buf.data(), &test_keys, sizeof(u16));
-	std::memcpy(input_buf.data() + 16, &coin_counter, sizeof(u16));
+	digital_input |= test_on ? 0x80 : 0x00;
+	std::memcpy(input_buf.data() + 128, &digital_input, sizeof(u64));
+	std::memcpy(input_buf.data() + 128 + 16, &coin_counter, sizeof(u16));
+
+	constexpr bool dip_switches[8] = {false, false, false, false, true, false, false, false};
+	for (int i = 0; i < 8; i++)
```
Comment (Megamouse) [inline rpcs3/Emu/Io/usio.cpp:373]:
```suggestion
	constexpr std::array<bool, 8> dip_switches = {false, false, false, false, true, false, false, false};
	for (usz i = 0; i < dip_switches.size(); i++)
```

File: rpcs3/Emu/Io/usio.cpp (hunk 2 above)
Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp:507]: To prevent spamming the log. 🙂
There are still some unimplemented read/write operations, but they don't affect any aspect of the games.

Comment (elad335) [reply rpcs3/Emu/Io/usio.cpp:507]: That's what `trace` level is for.

File: rpcs3/Emu/Io/usio_config.h [hunk 4]
```diff
@@ -8,18 +8,21 @@ enum class usio_btn
...
+	right,
+	taiko_hit_side_left,
+	taiko_hit_side_right,
+	taiko_hit_center_left,
+	taiko_hit_center_right,
+	tekken_button1,
+	tekken_button2,
+	tekken_button3,
+	tekken_button4,
+	tekken_button5,
```
Comment (Megamouse) [inline rpcs3/Emu/Io/usio_config.h:25]: if these's such a huge difference between the two, then why didn't you just make two configs ?
Isn't the needed config known before all the translators are called?

Comment (brian218) [reply rpcs3/Emu/Io/usio_config.h:25]: Pardon, do you mean this?
Image

Comment (Megamouse) [reply rpcs3/Emu/Io/usio_config.h:25]: Touché.
But at the time there wasn't a real separation between games, only a logical one.
Now that you actually change the function based on the game, you might as well add a config for each (if you want to)

File: rpcs3/Emu/Io/usio.cpp (hunk 2 above)
Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp]: Okay.

File: rpcs3/Emu/Io/usio.cpp (hunk 3 above)
Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp]: Fixed.

File: rpcs3/Emu/Io/usio.cpp (hunk 2 above)
Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp]: Okay.

File: rpcs3/Emu/Io/usio_config.h (hunk 4 above)
Comment (brian218) [reply rpcs3/Emu/Io/usio_config.h:25]: That's gonna be another work, it's not the aim of this PR though.
Currently it still looks fine visually as I only added 5 and removed 4.

File: rpcs3/Emu/Io/usio.cpp [hunk 5]
```diff
@@ -221,89 +222,162 @@ void usb_device_usio::translate_input()
...
-	test_keys |= test_on ? 0x80 : 0x00;
-	std::memcpy(input_buf.data(), &test_keys, sizeof(u16));
-	std::memcpy(input_buf.data() + 16, &coin_counter, sizeof(u16));
+	digital_input |= test_on ? 0x80 : 0x00;
+	std::memcpy(input_buf.data() + 128, &digital_input, sizeof(u64));
+	std::memcpy(input_buf.data() + 128 + 16, &coin_counter, sizeof(u16));
+
+	constexpr std::array<bool, 8> dip_switches {false, false, false, false, true, false, false, false};
+	for (usz i = 0; i < dip_switches.size(); i++)
+		input_buf[2] |= dip_switches[i] << i;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/usio.cpp:374]: you can actually do this whole thing as constexpr by just doing:
```suggestion
		constexpr u32 dip_switches = 0b00010000;
		input_buf[2] |= dip_switches;
```
or if you want to keep the code:
```suggestion
	constexpr u32 dip_switches = []()
	{
		constexpr std::array<bool, 8> dip_switches {false, false, false, false, true, false, false, false};
		u32 value = 0;
		for (usz i = 0; i < dip_switches.size(); i++)
			value |= dip_switches[i] << i;
		return value;
	}();
	input_buf[2] |= dip_switches;
```

Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp]: That's nice.

File: rpcs3/Emu/Io/usio.cpp [hunk 6]
```diff
@@ -221,89 +222,161 @@ void usb_device_usio::translate_input()
...
 
-	test_keys |= test_on ? 0x80 : 0x00;
-	std::memcpy(input_buf.data(), &test_keys, sizeof(u16));
-	std::memcpy(input_buf.data() + 16, &coin_counter, sizeof(u16));
+	digital_input |= test_on ? 0x80 : 0x00;
+	std::memcpy(input_buf.data() + 128, &digital_input, sizeof(u64));
+	std::memcpy(input_buf.data() + 128 + 16, &coin_counter, sizeof(u16));
+
+		constexpr u8 dip_switches = 0b00010000;
+		input_buf[2] |= dip_switches;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/usio.cpp:373]: Oops. I guess the suggestion had an indentation too many here.

Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp]: Fixed.

**Review by Megamouse — APPROVED** [review]

**Comment by brad86m** [comment]
rpcs3_2023_07_25_07_54_25_100

**Comment by brian218** [comment]
Unrelated issue.

> rpcs3_2023_07_25_07_54_25_100

Unrelated to this PR.

**Comment by brad86m** [comment]
Sorry, how is this unrelated when the issue is literally about support added for TTT2, yet I'm showing in my image that it does not work.

**Comment by brian218** [comment]
> Sorry, how is this unrelated when the issue is literally about support added for TTT2, yet I'm showing in my image that it does not work.

The aim of this PR was to add IO support for the games.
Your problem is due to game license check failing.

We cannot provide any support for game license check handling here.
